File: ts/web3js/src/instructions/ephemeral-spl-token-program/schedulePrivateTransfer.ts
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";

import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  DELEGATION_PROGRAM_ID,
  EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
  HYDRA_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
} from "../../constants.js";
import { encryptEd25519Recipient } from "./crypto.js";
import {
  deriveRentPda,
  deriveShuttleAta,
  deriveShuttleEphemeralAta,
  deriveVault,
} from "./ephemeralAta.js";
import { deriveTransferQueue } from "./transferQueue.js";

// ---------------------------------------------------------------------------
// Local constants (mirror the on-chain side)
// ---------------------------------------------------------------------------

const SCHEDULE_PRIVATE_TRANSFER_DISCRIMINATOR = 30;

const STASH_PDA_SEED = Buffer.from("stash");
const HYDRA_CRANK_SEED_PREFIX = Buffer.from("crank");
const BUFFER_SEED = Buffer.from("buffer");
const DELEGATION_RECORD_SEED = Buffer.from("delegation");
const DELEGATION_METADATA_SEED = Buffer.from("delegation-metadata");

// ---------------------------------------------------------------------------
// PDA derivation helpers
// ---------------------------------------------------------------------------

/**
 * Derive the swap-custody "stash" PDA for a `(user, mint)` pair.
 *
 * Seeds: `[b"stash", user, mint]` under `EPHEMERAL_SPL_TOKEN_PROGRAM_ID`.
 *
 * The stash PDA serves as both the owner of the destination ATA that
 * receives swap output and, at trigger time, the payer + owner signer
 * that the scheduled instruction self-CPIs into ix 25 with.
 */
export function deriveStashPda(
  user: PublicKey,
  mint: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [STASH_PDA_SEED, user.toBuffer(), mint.toBuffer()],
    EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
  );
}

/**
 * Derive the stash ATA — the SPL associated token account where the swap
 * output lands. Owner is the stash PDA.
 */
export function deriveStashAta(
  user: PublicKey,
  mint: PublicKey,
  tokenProgram: PublicKey = TOKEN_PROGRAM_ID,
): [PublicKey, number] {
  const [stashPda] = deriveStashPda(user, mint);
  return deriveAtaWithBump(stashPda, mint, tokenProgram);
}

/**
 * Derive the Hydra crank PDA for a schedule.
 *
 * The on-chain Hydra seed overwrites the first 4 bytes of `stashPda` with
 * `shuttle_id_le` so each `(user, mint, shuttleId)` triple gets its own
 * crank. Callers that only know the user + mint should derive `stashPda`
 * via `deriveStashPda` first.
 */
export function deriveHydraCrankPda(
  stashPda: PublicKey,
  shuttleId: number,
): [PublicKey, number] {
  if (
    !Number.isInteger(shuttleId) ||
    shuttleId < 0 ||
    shuttleId > 0xffff_ffff
  ) {
    throw new Error("shuttleId must fit in u32");
  }
  return PublicKey.findProgramAddressSync(
    [HYDRA_CRANK_SEED_PREFIX, hydraSeed(stashPda, shuttleId)],
    HYDRA_PROGRAM_ID,
  );
}

/**
 * Build the 32-byte Hydra crank seed for `(stashPda, shuttleId)`.
 *
 * Must match the on-chain `derive_hydra_seed` exactly: overwrite the
 * first 4 bytes of the stash PDA with `shuttle_id_le`. No hashing — we
 * don't need cryptographic uniformity, just a distinct 32-byte output
 * per tuple.
 */
function hydraSeed(stashPda: PublicKey, shuttleId: number): Buffer {
  const seed = Buffer.from(stashPda.toBuffer()); // 32-byte copy
  seed.writeUInt32LE(shuttleId, 0);
  return seed;
}

// ---------------------------------------------------------------------------
// Instruction builder
// ---------------------------------------------------------------------------

/**
 * Build a `schedule_private_transfer` instruction (discriminator 30).
 *
 * Appends to a swap transaction so the swap's `destinationTokenAccount`
 * (which the caller sets to the stash ATA) gets privately forwarded to
 * `destinationOwner` after the scheduled ix 25 fires via Hydra.
 *
 * Accounts passed (7):
 *   0 user (signer, writable — funds the stash PDA)
 *   1 stash_pda (writable)
 *   2 rent_pda (writable — sponsors Hydra crank rent + CRANKER_REWARD)
 *   3 hydra_crank_pda (writable — created by the Hydra CPI)
 *   4 hydra_program
 *   5 system_program
 *   6 token_program
 *
 * All 14 pubkeys that ix 25 will consume at trigger time are derived
 * on-chain from 10 client-supplied bumps (included in the instruction
 * data), so the outer tx stays compact.
 *
 * @param user The user who owns the stash PDA and signs the outer tx.
 * @param mint The output-mint of the swap.
 * @param shuttleId A u32 identifier for the shuttle (client-chosen).
 * @param destinationOwner The wallet that will ultimately receive the
 *   private transfer. Encrypted to `validator` before going on-chain.
 * @param minDelayMs Earliest the queued transfer may settle.
 * @param maxDelayMs Latest it may settle.
 * @param split Number of queue entries to split the transfer across.
 * @param validator The validator that owns the transfer-queue PDA.
 * @param tokenProgram Override for the SPL token program (defaults to
 *   classic Token).
 * @param clientRefId Optional u64 correlation id attached to each split.
 */
export function schedulePrivateTransferIx(
  user: PublicKey,
  mint: PublicKey,
  shuttleId: number,
  destinationOwner: PublicKey,
  minDelayMs: bigint,
  maxDelayMs: bigint,
  split: number,
  validator: PublicKey,
  tokenProgram: PublicKey = TOKEN_PROGRAM_ID,
  clientRefId?: bigint,
): TransactionInstruction {
  if (
    !Number.isInteger(shuttleId) ||
    shuttleId < 0 ||
    shuttleId > 0xffff_ffff
  ) {
    throw new Error("shuttleId must fit in u32");
  }
  if (!Number.isInteger(split) || split <= 0 || split > 0xffff_ffff) {
    throw new Error("split must be a positive u32");
  }
  if (
    minDelayMs < 0n ||
    maxDelayMs < 0n ||
    (clientRefId !== undefined && clientRefId < 0n)
  ) {
    throw new Error("delays and clientRefId must be non-negative");
  }
  if (maxDelayMs < minDelayMs) {
    throw new Error("maxDelayMs must be greater than or equal to minDelayMs");
  }
  const U64_MAX = 0xffff_ffff_ffff_ffffn;
  if (
    minDelayMs > U64_MAX ||
    maxDelayMs > U64_MAX ||
    (clientRefId !== undefined && clientRefId > U64_MAX)
  ) {
    throw new Error("delays and clientRefId must fit in u64");
  }

  // -------- derive every pubkey + bump ix 25 will need --------
  const [stashPda, stashBump] = deriveStashPda(user, mint);
  const [, stashAtaBump] = deriveAtaWithBump(stashPda, mint, tokenProgram);
  const [rentPda] = deriveRentPda();
  const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(
    stashPda,
    mint,
    shuttleId,
  );
  const [shuttleAta, shuttleEataBump] = deriveShuttleAta(
    shuttleEphemeralAta,
    mint,
  );
  const [, shuttleWalletAtaBump] = deriveAtaWithBump(
    shuttleEphemeralAta,
    mint,
    tokenProgram,
  );
  const [, bufferBump] = PublicKey.findProgramAddressSync(
    [BUFFER_SEED, shuttleAta.toBuffer()],
    EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
  );
  const [, delegationRecordBump] = PublicKey.findProgramAddressSync(
    [DELEGATION_RECORD_SEED, shuttleAta.toBuffer()],
    DELEGATION_PROGRAM_ID,
  );
  const [, delegationMetadataBump] = PublicKey.findProgramAddressSync(
    [DELEGATION_METADATA_SEED, shuttleAta.toBuffer()],
    DELEGATION_PROGRAM_ID,
  );
  const [vault, globalVaultBump] = deriveVault(mint);
  const [, vaultTokenBump] = deriveAtaWithBump(vault, mint, tokenProgram);
  const [, queueBump] = deriveTransferQueue(mint, validator);
  const [hydraCrankPda] = deriveHydraCrankPda(stashPda, shuttleId);

  // -------- build the encrypted payload (same shape as ix 25) --------
  const encryptedDestination = encryptEd25519Recipient(
    destinationOwner.toBytes(),
    validator,
  );
  const encryptedSuffix = encryptEd25519Recipient(
    packPrivateTransferSuffix(minDelayMs, maxDelayMs, split, clientRefId),
    validator,
  );

  // -------- wire data --------
  // Wire: [disc][shuttle_id][stash_bump][mint(32)][10 bumps][3 vardata blobs]
  // Offsets below mirror the on-chain processor's FIXED_PREFIX_LEN = 47
  // (shuttle_id..queue_bump) followed by validator/enc_dest/enc_suffix.
  const data = Buffer.concat([
    Buffer.from([SCHEDULE_PRIVATE_TRANSFER_DISCRIMINATOR]),
    u32leBuffer(shuttleId),
    Buffer.from([stashBump]),
    mint.toBuffer(),
    Buffer.from([shuttleBump]),
    Buffer.from([shuttleEataBump]),
    Buffer.from([shuttleWalletAtaBump]),
    Buffer.from([bufferBump]),
    Buffer.from([delegationRecordBump]),
    Buffer.from([delegationMetadataBump]),
    Buffer.from([globalVaultBump]),
    Buffer.from([vaultTokenBump]),
    Buffer.from([stashAtaBump]),
    Buffer.from([queueBump]),
    encodeLengthPrefixedBytes(validator.toBytes()),
    encodeLengthPrefixedBytes(encryptedDestination),
    encodeLengthPrefixedBytes(encryptedSuffix),
  ]);

  return new TransactionInstruction({
    programId: EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
    keys: [
      { pubkey: user, isSigner: true, isWritable: true },
      { pubkey: stashPda, isSigner: false, isWritable: true },
      { pubkey: rentPda, isSigner: false, isWritable: true },
      { pubkey: hydraCrankPda, isSigner: false, isWritable: true },
      { pubkey: HYDRA_PROGRAM_ID, isSigner: false, isWritable: false },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
      { pubkey: tokenProgram, isSigner: false, isWritable: false },
    ],
    data,
  });
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function deriveAtaWithBump(
  wallet: PublicKey,
  mint: PublicKey,
  tokenProgram: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [wallet.toBuffer(), tokenProgram.toBuffer(), mint.toBuffer()],
    ASSOCIATED_TOKEN_PROGRAM_ID,
  );
}

function encodeLengthPrefixedBytes(bytes: Uint8Array): Buffer {
  if (bytes.length > 0xff) {
    throw new Error("payload exceeds u8 length");
  }
  return Buffer.concat([Buffer.from([bytes.length]), Buffer.from(bytes)]);
}

function packPrivateTransferSuffix(
  minDelayMs: bigint,
  maxDelayMs: bigint,
  split: number,
  clientRefId?: bigint,
): Buffer {
  const suffix = Buffer.alloc(
    clientRefId === undefined ? 8 + 8 + 4 : 8 + 8 + 4 + 8,
  );
  suffix.writeBigUInt64LE(minDelayMs, 0);
  suffix.writeBigUInt64LE(maxDelayMs, 8);
  suffix.writeUInt32LE(split, 16);
  if (clientRefId !== undefined) {
    suffix.writeBigUInt64LE(clientRefId, 20);
  }
  return suffix;
}

function u32leBuffer(value: number): Buffer {
  const out = Buffer.alloc(4);
  out.writeUInt32LE(value, 0);
  return out;
}

File: rust/sdk/src/spl/builders/schedule_private_transfer.rs
use core::fmt;

use crate::{
    compat,
    consts::{ESPL_TOKEN_PROGRAM_ID, HYDRA_PROGRAM_ID},
    cpi::DELEGATION_PROGRAM_ID,
    spl::{
        find_hydra_crank_pda, find_rent_pda, find_shuttle_ata, find_shuttle_ephemeral_ata,
        find_stash_ata, find_stash_pda, find_transfer_queue,
        types::find_associated_token_address_with_bump, EphemeralSplDiscriminator, GlobalVault,
    },
};

const BUFFER_SEED: &[u8] = b"buffer";
const DELEGATION_RECORD_SEED: &[u8] = b"delegation";
const DELEGATION_METADATA_SEED: &[u8] = b"delegation-metadata";

#[derive(Debug)]
pub enum SchedulePrivateTransferBuilderError {
    InvalidSplit(u32),
    InvalidDelayRange {
        min_delay_ms: u64,
        max_delay_ms: u64,
    },
    PrivateTransferPayloadTooLong(usize),
    Encryption(dlp_api::encryption::EncryptionError),
}

impl fmt::Display for SchedulePrivateTransferBuilderError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidSplit(split) => write!(f, "split must be greater than zero, got {split}"),
            Self::InvalidDelayRange {
                min_delay_ms,
                max_delay_ms,
            } => write!(
                f,
                "max_delay_ms ({max_delay_ms}) must be greater than or equal to min_delay_ms ({min_delay_ms})"
            ),
            Self::PrivateTransferPayloadTooLong(len) => {
                write!(f, "encrypted private transfer payload exceeds u8 length: {len}")
            }
            Self::Encryption(err) => write!(f, "private transfer encryption failed: {err}"),
        }
    }
}

pub struct SchedulePrivateTransferBuilder {
    pub user: compat::Pubkey,
    pub mint: compat::Pubkey,
    pub shuttle_id: u32,
    pub destination_owner: compat::Pubkey,
    pub min_delay_ms: u64,
    pub max_delay_ms: u64,
    pub split: u32,
    pub validator: compat::Pubkey,
    pub token_program: compat::Pubkey,
    pub client_ref_id: Option<u64>,
}

impl SchedulePrivateTransferBuilder {
    #[inline(always)]
    pub fn instruction(&self) -> Result<compat::Instruction, SchedulePrivateTransferBuilderError> {
        if self.split == 0 {
            return Err(SchedulePrivateTransferBuilderError::InvalidSplit(
                self.split,
            ));
        }
        if self.max_delay_ms < self.min_delay_ms {
            return Err(SchedulePrivateTransferBuilderError::InvalidDelayRange {
                min_delay_ms: self.min_delay_ms,
                max_delay_ms: self.max_delay_ms,
            });
        }

        let (stash_pda, stash_bump) = find_stash_pda(&self.user, &self.mint);
        let (_stash_ata, stash_ata_bump) =
            find_stash_ata(&self.user, &self.mint, &self.token_program);
        let (rent_pda, _rent_bump) = find_rent_pda();
        let (shuttle_ephemeral_ata, shuttle_bump) =
            find_shuttle_ephemeral_ata(&stash_pda, &self.mint, self.shuttle_id);
        let (shuttle_ata, shuttle_ata_bump) = find_shuttle_ata(&shuttle_ephemeral_ata, &self.mint);
        let (_shuttle_wallet_ata, shuttle_wallet_ata_bump) =
            find_associated_token_address_with_bump(
                &shuttle_ephemeral_ata,
                &self.mint,
                &self.token_program,
            );
        let (_buffer, buffer_bump) = compat::Pubkey::find_program_address(
            &[BUFFER_SEED, shuttle_ata.as_ref()],
            &ESPL_TOKEN_PROGRAM_ID,
        );
        let (_delegation_record, delegation_record_bump) = compat::Pubkey::find_program_address(
            &[DELEGATION_RECORD_SEED, shuttle_ata.as_ref()],
            &DELEGATION_PROGRAM_ID,
        );
        let (_delegation_metadata, delegation_metadata_bump) = compat::Pubkey::find_program_address(
            &[DELEGATION_METADATA_SEED, shuttle_ata.as_ref()],
            &DELEGATION_PROGRAM_ID,
        );
        let (vault, global_vault_bump) = GlobalVault::find_pda(&self.mint);
        let (_vault_ata, vault_token_bump) =
            find_associated_token_address_with_bump(&vault, &self.mint, &self.token_program);
        let (_queue, queue_bump) = find_transfer_queue(&self.mint, &self.validator);
        let (hydra_crank_pda, _hydra_crank_bump) =
            find_hydra_crank_pda(&stash_pda, self.shuttle_id);

        let encrypted_destination =
            encrypt_private_transfer_field(self.destination_owner.as_ref(), &self.validator)?;
        let encrypted_suffix = encrypt_private_transfer_field(
            &pack_private_transfer_suffix(
                self.min_delay_ms,
                self.max_delay_ms,
                self.split,
                self.client_ref_id,
            ),
            &self.validator,
        )?;

        let mut data = Vec::with_capacity(
            1 + 4
                + 1
                + 32
                + 10
                + 1
                + 32
                + 1
                + encrypted_destination.len()
                + 1
                + encrypted_suffix.len(),
        );
        data.push(EphemeralSplDiscriminator::SchedulePrivateTransfer as u8);
        data.extend_from_slice(&self.shuttle_id.to_le_bytes());
        data.push(stash_bump);
        data.extend_from_slice(self.mint.as_ref());
        data.push(shuttle_bump);
        data.push(shuttle_ata_bump);
        data.push(shuttle_wallet_ata_bump);
        data.push(buffer_bump);
        data.push(delegation_record_bump);
        data.push(delegation_metadata_bump);
        data.push(global_vault_bump);
        data.push(vault_token_bump);
        data.push(stash_ata_bump);
        data.push(queue_bump);
        push_length_prefixed(&mut data, self.validator.as_ref())?;
        push_length_prefixed(&mut data, &encrypted_destination)?;
        push_length_prefixed(&mut data, &encrypted_suffix)?;

        Ok(compat::Instruction {
            program_id: ESPL_TOKEN_PROGRAM_ID,
            accounts: vec![
                compat::AccountMeta::new(self.user, true),
                compat::AccountMeta::new(stash_pda, false),
                compat::AccountMeta::new(rent_pda, false),
                compat::AccountMeta::new(hydra_crank_pda, false),
                compat::AccountMeta::new_readonly(HYDRA_PROGRAM_ID, false),
                compat::AccountMeta::new_readonly(
                    compat::Pubkey::new_from_array(solana_system_interface::program::ID.to_bytes()),
                    false,
                ),
                compat::AccountMeta::new_readonly(self.token_program, false),
            ],
            data,
        })
    }
}

#[inline(always)]
fn pack_private_transfer_suffix(
    min_delay_ms: u64,
    max_delay_ms: u64,
    split: u32,
    client_ref_id: Option<u64>,
) -> Vec<u8> {
    let mut suffix = Vec::with_capacity(if client_ref_id.is_some() { 28 } else { 20 });
    suffix.extend_from_slice(&min_delay_ms.to_le_bytes());
    suffix.extend_from_slice(&max_delay_ms.to_le_bytes());
    suffix.extend_from_slice(&split.to_le_bytes());
    if let Some(client_ref_id) = client_ref_id {
        suffix.extend_from_slice(&client_ref_id.to_le_bytes());
    }
    suffix
}

#[inline(always)]
fn push_length_prefixed(
    data: &mut Vec<u8>,
    bytes: &[u8],
) -> Result<(), SchedulePrivateTransferBuilderError> {
    let len = u8::try_from(bytes.len()).map_err(|_| {
        SchedulePrivateTransferBuilderError::PrivateTransferPayloadTooLong(bytes.len())
    })?;
    data.push(len);
    data.extend_from_slice(bytes);
    Ok(())
}

#[inline(always)]
fn encrypt_private_transfer_field(
    plaintext: &[u8],
    validator: &compat::Pubkey,
) -> Result<Vec<u8>, SchedulePrivateTransferBuilderError> {
    dlp_api::encryption::encrypt_ed25519_recipient(plaintext, validator.as_array())
        .map_err(SchedulePrivateTransferBuilderError::Encryption)
}
